### PR TITLE
workflow: restart if processing stale data INSPIR-1854

### PR DESCRIPTION
**Pull request by ammirate**

### CODE REVIEW

File: inspirehep/modules/workflows/tasks/actions.py [hunk 1]
```diff
@@ -56,6 +56,7 @@
     extract_references_from_text,
 )
 from inspirehep.modules.refextract.matcher import match_references
+from inspirehep.modules.workflows.config import WORKFLOWS_RESTART_LIMIT
```
Comment (drjova) [inline inspirehep/modules/workflows/tasks/actions.py]: I'd put it on the main ``config.py`` file so we can easily change it

File: inspirehep/modules/workflows/tasks/actions.py [hunk 2]
```diff
@@ -629,3 +631,37 @@ def load_from_source_data(obj, eng):
...
+    # XXX: for some reasons, restarting a workflow generates a fork. The
+    # current branch stops here, the other one will start from the beginning.
+    eng.stop()
+
+
+@with_debug_logging
+def increase_restart_count_or_error(obj, eng):
+    """Increase the `restart-count` of the current workflow.
+
+    If `restart-count` contains a value greater than 3 it throws a WorkflowError.
```
Comment (drjova) [inline inspirehep/modules/workflows/tasks/actions.py]:
```suggestion
    If `restart-count` contains a value greater than ``WORKFLOWS_RESTART_LIMIT`` it throws a WorkflowError.
```

File: inspirehep/modules/workflows/tasks/upload.py [hunk 3]
```diff
@@ -120,3 +120,29 @@ def set_schema(obj, eng):
         obj.log.debug('Schema already is url')
 
     obj.log.debug('Final schema %s', obj.data['$schema'])
+
+
+def _is_stale_data(workflow_object):
+    if not workflow_object.extra_data.get('is-update') or \
```
[the remark this answers is not in the document]
Comment (drjova) [reply inspirehep/modules/workflows/tasks/upload.py]: is the ``head_version_id`` the same with ``head_uuid``?

File: inspirehep/modules/workflows/tasks/actions.py (hunk 2 above)
Comment (ammirate) [reply inspirehep/modules/workflows/tasks/actions.py]: Ops!

File: inspirehep/modules/workflows/tasks/actions.py [hunk 4]
```diff
@@ -629,3 +632,37 @@ def load_from_source_data(obj, eng):
...
+    `extra_data.persistent_data.marks` the increment is skipped.
+    """
+    try:
+        count = obj.extra_data['source_data']['persistent_data']['marks']['restart-count']
+    except KeyError:
+        count = None
+
+    if count is None:
+        return
+    if count < WORKFLOWS_RESTART_LIMIT:
```
Comment (drjova) [inline inspirehep/modules/workflows/tasks/actions.py]: ``current_app.get('WORKFLOWS_RESTART_LIMIT')`` 😄

File: inspirehep/modules/workflows/tasks/upload.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (drjova) [reply inspirehep/modules/workflows/tasks/upload.py]: @ammirate I'm just wondering about that

[the remark this answers is not in the document]
Comment (ammirate) [reply inspirehep/modules/workflows/tasks/upload.py]: yep

File: tests/integration/workflows/test_arxiv_workflow.py [hunk 5]
```diff
@@ -1057,12 +1057,13 @@ def _return_value(obj, eng):
...
         'two': {'id': 2}
-    }
+    }.viewitems()
+
     expected_persistent_data_second_run = {
         'one': {'id': 1},
         'two': {'id': 2},
         'three': {'id': 43},
-    }
+    }.viewitems()
```
Comment (drjova) [inline tests/integration/workflows/test_arxiv_workflow.py]: ```six.viewitems``?

Comment (ammirate) [reply tests/integration/workflows/test_arxiv_workflow.py]: Do we want to make workflows and workflows tests ready for python3?

**Pull request merged**